File: python/mmwcore/tracking/metrics.py
"""Sequence-level summaries for tracker validation."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass

import numpy as np

from mmwcore import _native
from mmwcore.core import ScenerySpec, TrackFrame, TrackStatus

_STATUS_CODES = {
    TrackStatus.TENTATIVE: 0,
    TrackStatus.CONFIRMED: 1,
    TrackStatus.COASTING: 2,
}


@dataclass(frozen=True)
class TrackObservationSummary:
    """Observed lifetime of one track ID within a sequence."""

    track_id: int
    observed_frames: int
    confirmed_frames: int
    first_frame_index: int
    last_frame_index: int
    first_position_m: tuple[float, float, float]
    last_position_m: tuple[float, float, float]
    median_position_m: tuple[float, float, float]
    displacement_m: float
    path_length_m: float
    median_speed_mps: float
    max_speed_mps: float
    confirmed_intervals: tuple[tuple[int, int], ...]
    in_scenery_frames: int | None
    outside_scenery_frames: int | None


@dataclass(frozen=True)
class TrackingSequenceSummary:
    """Identity and coverage statistics over an ordered frame sequence."""

    num_frames: int
    frames_with_tracks: int
    frames_with_confirmed_tracks: int
    max_concurrent_tracks: int
    tracks: tuple[TrackObservationSummary, ...]

    @property
    def observed_track_ids(self) -> int:
        return len(self.tracks)

    @property
    def confirmed_track_ids(self) -> int:
        return sum(track.confirmed_frames > 0 for track in self.tracks)

    @property
    def confirmed_frame_coverage(self) -> float:
        if self.num_frames == 0:
            return 0.0
        return self.frames_with_confirmed_tracks / self.num_frames

    def to_record(self) -> dict[str, object]:
        return {
            "num_frames": self.num_frames,
            "frames_with_tracks": self.frames_with_tracks,
            "frames_with_confirmed_tracks": self.frames_with_confirmed_tracks,
            "confirmed_frame_coverage": self.confirmed_frame_coverage,
            "max_concurrent_tracks": self.max_concurrent_tracks,
            "observed_track_ids": self.observed_track_ids,
            "confirmed_track_ids": self.confirmed_track_ids,
            "tracks": [
                {
                    "track_id": track.track_id,
                    "observed_frames": track.observed_frames,
                    "confirmed_frames": track.confirmed_frames,
                    "first_frame_index": track.first_frame_index,
                    "last_frame_index": track.last_frame_index,
                    "first_position_m": list(track.first_position_m),
                    "last_position_m": list(track.last_position_m),
                    "median_position_m": list(track.median_position_m),
                    "displacement_m": track.displacement_m,
                    "path_length_m": track.path_length_m,
                    "median_speed_mps": track.median_speed_mps,
                    "max_speed_mps": track.max_speed_mps,
                    "confirmed_intervals": [
                        [start, stop] for start, stop in track.confirmed_intervals
                    ],
                    "in_scenery_frames": track.in_scenery_frames,
                    "outside_scenery_frames": track.outside_scenery_frames,
                }
                for track in self.tracks
            ],
        }


def summarize_track_frames(
    frames: Iterable[TrackFrame],
    *,
    scenery: ScenerySpec | None = None,
    frame_index_offset: int = 0,
) -> TrackingSequenceSummary:
    """Summarize ordered tracker reports without assuming ground-truth identities."""

    if frame_index_offset < 0:
        raise ValueError("Tracking summary frame_index_offset must be non-negative.")
    packed = _pack_track_frames(tuple(frames))
    scenery_boxes = (
        [(box.x_min_m, box.x_max_m, box.y_min_m, box.y_max_m) for box in scenery.boundary_boxes]
        if scenery is not None
        else None
    )
    result = _native.summarize_tracking_metrics(
        packed,
        scenery_boxes,
        frame_index_offset,
    )
    header, identity, motion, interval_data = result
    (
        num_frames,
        frames_with_tracks,
        frames_with_confirmed,
        max_concurrent,
    ) = header
    (
        track_ids,
        observed_frames,
        confirmed_frames,
        first_frame_indices,
        last_frame_indices,
        first_positions,
        last_positions,
        median_positions,
    ) = identity
    displacement, path_length, median_speed, max_speed = motion
    interval_offsets, intervals, in_scenery_frames, outside_scenery_frames = interval_data
    tracks = tuple(
        _track_summary_from_native(
            index,
            track_ids=track_ids,
            observed_frames=observed_frames,
            confirmed_frames=confirmed_frames,
            first_frame_indices=first_frame_indices,
            last_frame_indices=last_frame_indices,
            first_positions=first_positions,
            last_positions=last_positions,
            median_positions=median_positions,
            displacement=displacement,
            path_length=path_length,
            median_speed=median_speed,
            max_speed=max_speed,
            interval_offsets=interval_offsets,
            intervals=intervals,
            in_scenery_frames=in_scenery_frames,
            outside_scenery_frames=outside_scenery_frames,
        )
        for index in range(track_ids.size)
    )
    return TrackingSequenceSummary(
        num_frames=num_frames,
        frames_with_tracks=frames_with_tracks,
        frames_with_confirmed_tracks=frames_with_confirmed,
        max_concurrent_tracks=max_concurrent,
        tracks=tracks,
    )


def _pack_track_frames(
    frames: tuple[TrackFrame, ...],
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    observation_count = sum(frame.num_tracks for frame in frames)
    frame_offsets = np.empty(len(frames) + 1, dtype=np.int64)
    track_ids = np.empty(observation_count, dtype=np.int64)
    positions = np.empty((observation_count, 3), dtype=np.float32)
    velocities = np.empty((observation_count, 3), dtype=np.float32)
    status_codes = np.empty(observation_count, dtype=np.uint8)
    frame_offsets[0] = 0
    cursor = 0
    for frame_index, frame in enumerate(frames):
        stop = cursor + frame.num_tracks
        track_ids[cursor:stop] = frame.track_ids
        positions[cursor:stop] = frame.positions
        velocities[cursor:stop] = frame.velocities
        status_codes[cursor:stop] = [_STATUS_CODES[status] for status in frame.statuses]
        frame_offsets[frame_index + 1] = stop
        cursor = stop
    return (
        frame_offsets,
        track_ids,
        positions,
        velocities,
        status_codes,
    )


def _track_summary_from_native(
    index: int,
    *,
    track_ids: np.ndarray,
    observed_frames: np.ndarray,
    confirmed_frames: np.ndarray,
    first_frame_indices: np.ndarray,
    last_frame_indices: np.ndarray,
    first_positions: np.ndarray,
    last_positions: np.ndarray,
    median_positions: np.ndarray,
    displacement: np.ndarray,
    path_length: np.ndarray,
    median_speed: np.ndarray,
    max_speed: np.ndarray,
    interval_offsets: np.ndarray,
    intervals: np.ndarray,
    in_scenery_frames: np.ndarray,
    outside_scenery_frames: np.ndarray,
) -> TrackObservationSummary:
    interval_start = int(interval_offsets[index])
    interval_stop = int(interval_offsets[index + 1])
    in_scenery = int(in_scenery_frames[index])
    outside_scenery = int(outside_scenery_frames[index])
    return TrackObservationSummary(
        track_id=int(track_ids[index]),
        observed_frames=int(observed_frames[index]),
        confirmed_frames=int(confirmed_frames[index]),
        first_frame_index=int(first_frame_indices[index]),
        last_frame_index=int(last_frame_indices[index]),
        first_position_m=_vector3(first_positions[index]),
        last_position_m=_vector3(last_positions[index]),
        median_position_m=_vector3(median_positions[index]),
        displacement_m=float(displacement[index]),
        path_length_m=float(path_length[index]),
        median_speed_mps=float(median_speed[index]),
        max_speed_mps=float(max_speed[index]),
        confirmed_intervals=tuple(
            (int(start), int(stop)) for start, stop in intervals[interval_start:interval_stop]
        ),
        in_scenery_frames=None if in_scenery < 0 else in_scenery,
        outside_scenery_frames=None if outside_scenery < 0 else outside_scenery,
    )


def _vector3(values: np.ndarray) -> tuple[float, float, float]:
    return (float(values[0]), float(values[1]), float(values[2]))

File: crates/mmwcore/src/tracking/metrics.rs
//! Native sequence-level tracking metrics.

use std::collections::{BTreeMap, HashSet};
use std::fmt;

use super::Box2D;

/// Packed observations from ordered tracker frames.
#[derive(Clone, Copy, Debug)]
pub struct TrackingMetricsInput<'a> {
    /// Cumulative observation counts, with one entry per frame plus the final count.
    pub frame_offsets: &'a [usize],
    pub track_ids: &'a [i64],
    /// Packed XYZ positions with shape `(observations, 3)`.
    pub positions: &'a [f32],
    /// Packed XYZ velocities with shape `(observations, 3)`.
    pub velocities: &'a [f32],
    /// Native lifecycle status codes aligned with observations.
    pub status_codes: &'a [u8],
    /// `None` means scenery metrics are disabled; `Some([])` means every point is in scenery.
    pub scenery_boxes: Option<&'a [Box2D]>,
    pub frame_index_offset: usize,
}

/// One track's observed motion and lifecycle summary.
#[derive(Clone, Debug, PartialEq)]
pub struct TrackObservationMetrics {
    pub track_id: i64,
    pub observed_frames: usize,
    pub confirmed_frames: usize,
    pub first_frame_index: usize,
    pub last_frame_index: usize,
    pub first_position_m: [f32; 3],
    pub last_position_m: [f32; 3],
    pub median_position_m: [f32; 3],
    pub displacement_m: f32,
    pub path_length_m: f32,
    pub median_speed_mps: f32,
    pub max_speed_mps: f32,
    pub confirmed_intervals: Vec<[usize; 2]>,
    pub in_scenery_frames: Option<usize>,
    pub outside_scenery_frames: Option<usize>,
}

/// Aggregated tracker coverage over an ordered frame sequence.
#[derive(Clone, Debug, PartialEq)]
pub struct TrackingSequenceMetrics {
    pub num_frames: usize,
    pub frames_with_tracks: usize,
    pub frames_with_confirmed_tracks: usize,
    pub max_concurrent_tracks: usize,
    pub tracks: Vec<TrackObservationMetrics>,
}

/// Summarize ordered tracker reports in a single native call.
pub fn summarize_tracking_metrics(
    input: TrackingMetricsInput<'_>,
) -> Result<TrackingSequenceMetrics, TrackingMetricsError> {
    let num_frames = validate_input(input)?;
    let mut observed = BTreeMap::<i64, Vec<Observation>>::new();
    let mut frames_with_tracks = 0_usize;
    let mut frames_with_confirmed_tracks = 0_usize;
    let mut max_concurrent_tracks = 0_usize;

    for relative_frame_index in 0..num_frames {
        let start = input.frame_offsets[relative_frame_index];
        let stop = input.frame_offsets[relative_frame_index + 1];
        let track_count = stop - start;
        frames_with_tracks += usize::from(track_count > 0);
        max_concurrent_tracks = max_concurrent_tracks.max(track_count);
        let frame_index = input
            .frame_index_offset
            .checked_add(relative_frame_index)
            .ok_or(TrackingMetricsError::FrameIndexOverflow)?;
        let mut seen_track_ids = HashSet::with_capacity(track_count);
        let mut confirmed_in_frame = false;
        for observation_index in start..stop {
            let track_id = input.track_ids[observation_index];
            if !seen_track_ids.insert(track_id) {
                return Err(TrackingMetricsError::DuplicateTrackId {
                    frame_index,
                    track_id,
                });
            }
            let status_code = input.status_codes[observation_index];
            let confirmed = status_code == 1;
            confirmed_in_frame |= confirmed;
            observed.entry(track_id).or_default().push(Observation {
                frame_index,
                position: vector3(input.positions, observation_index),
                velocity: vector3(input.velocities, observation_index),
                confirmed,
            });
        }
        frames_with_confirmed_tracks += usize::from(confirmed_in_frame);
    }

    let tracks = observed
        .into_iter()
        .map(|(track_id, observations)| {
            summarize_track(track_id, &observations, input.scenery_boxes)
        })
        .collect();
    Ok(TrackingSequenceMetrics {
        num_frames,
        frames_with_tracks,
        frames_with_confirmed_tracks,
        max_concurrent_tracks,
        tracks,
    })
}

#[derive(Clone, Copy, Debug)]
struct Observation {
    frame_index: usize,
    position: [f32; 3],
    velocity: [f32; 3],
    confirmed: bool,
}

fn validate_input(input: TrackingMetricsInput<'_>) -> Result<usize, TrackingMetricsError> {
    let Some((&first_offset, remaining_offsets)) = input.frame_offsets.split_first() else {
        return Err(TrackingMetricsError::EmptyFrameOffsets);
    };
    if first_offset != 0 {
        return Err(TrackingMetricsError::InvalidFirstFrameOffset {
            actual: first_offset,
        });
    }
    let observation_count = input.track_ids.len();
    for (index, &offset) in remaining_offsets.iter().enumerate() {
        let previous = input.frame_offsets[index];
        if offset < previous {
            return Err(TrackingMetricsError::NonMonotonicFrameOffset {
                index: index + 1,
                previous,
                actual: offset,
            });
        }
        if offset > observation_count {
            return Err(TrackingMetricsError::FrameOffsetOutOfBounds {
                index: index + 1,
                value: offset,
                observation_count,
            });
        }
    }
    let final_offset = *input.frame_offsets.last().expect("non-empty is checked");
    if final_offset != observation_count {
        return Err(TrackingMetricsError::FinalFrameOffset {
            expected: observation_count,
            actual: final_offset,
        });
    }
    validate_vector_length("positions", input.positions.len(), observation_count)?;
    validate_vector_length("velocities", input.velocities.len(), observation_count)?;
    if input.status_codes.len() != observation_count {
        return Err(TrackingMetricsError::InputLength {
            name: "status codes",
            expected: observation_count,
            actual: input.status_codes.len(),
        });
    }
    for (index, &track_id) in input.track_ids.iter().enumerate() {
        if track_id < 0 {
            return Err(TrackingMetricsError::NegativeTrackId { index, track_id });
        }
    }
    for (index, &status_code) in input.status_codes.iter().enumerate() {
        if status_code > 2 {
            return Err(TrackingMetricsError::InvalidStatusCode { index, status_code });
        }
    }
    for (name, values) in [
        ("positions", input.positions),
        ("velocities", input.velocities),
    ] {
        if let Some((index, _)) = values
            .iter()
            .enumerate()
            .find(|(_, value)| !value.is_finite())
        {
            return Err(TrackingMetricsError::NonFiniteInput { name, index });
        }
    }
    Ok(input.frame_offsets.len() - 1)
}

fn validate_vector_length(
    name: &'static str,
    actual: usize,
    observation_count: usize,
) -> Result<(), TrackingMetricsError> {
    let expected = observation_count
        .checked_mul(3)
        .ok_or(TrackingMetricsError::InputLengthOverflow { name })?;
    if actual != expected {
        return Err(TrackingMetricsError::InputLength {
            name,
            expected,
            actual,
        });
    }
    Ok(())
}

fn vector3(values: &[f32], index: usize) -> [f32; 3] {
    let offset = index * 3;
    [values[offset], values[offset + 1], values[offset + 2]]
}

fn summarize_track(
    track_id: i64,
    observations: &[Observation],
    scenery_boxes: Option<&[Box2D]>,
) -> TrackObservationMetrics {
    let first = observations
        .first()
        .expect("tracked observations are non-empty");
    let last = observations
        .last()
        .expect("tracked observations are non-empty");
    let confirmed_frames = observations
        .iter()
        .filter_map(|observation| observation.confirmed.then_some(observation.frame_index))
        .collect::<Vec<_>>();
    let median_position_m = [
        median(
            observations
                .iter()
                .map(|observation| observation.position[0])
                .collect(),
        ),
        median(
            observations
                .iter()
                .map(|observation| observation.position[1])
                .collect(),
        ),
        median(
            observations
                .iter()
                .map(|observation| observation.position[2])
                .collect(),
        ),
    ];
    let speeds = observations
        .iter()
        .map(|observation| observation.velocity[0].hypot(observation.velocity[1]))
        .collect::<Vec<_>>();
    let path_length_m = observations
        .windows(2)
        .map(|window| {
            (window[1].position[0] - window[0].position[0])
                .hypot(window[1].position[1] - window[0].position[1])
        })
        .sum();
    let in_scenery_frames = scenery_boxes.map(|boxes| {
        observations
            .iter()
            .filter(|observation| scenery_contains(boxes, observation.position))
            .count()
    });
    TrackObservationMetrics {
        track_id,
        observed_frames: observations.len(),
        confirmed_frames: confirmed_frames.len(),
        first_frame_index: first.frame_index,
        last_frame_index: last.frame_index,
        first_position_m: first.position,
        last_position_m: last.position,
        median_position_m,
        displacement_m: (last.position[0] - first.position[0])
            .hypot(last.position[1] - first.position[1]),
        path_length_m,
        median_speed_mps: median(speeds.clone()),
        max_speed_mps: speeds.into_iter().fold(f32::NEG_INFINITY, f32::max),
        confirmed_intervals: contiguous_intervals(&confirmed_frames),
        in_scenery_frames,
        outside_scenery_frames: in_scenery_frames.map(|count| observations.len() - count),
    }
}

fn median(mut values: Vec<f32>) -> f32 {
    values.sort_by(f32::total_cmp);
    let middle = values.len() / 2;
    if values.len() % 2 == 1 {
        values[middle]
    } else {
        (values[middle - 1] + values[middle]) * 0.5
    }
}

fn scenery_contains(boxes: &[Box2D], position: [f32; 3]) -> bool {
    boxes.is_empty()
        || boxes
            .iter()
            .copied()
            .any(|boundary| boundary.contains(f64::from(position[0]), f64::from(position[1])))
}

fn contiguous_intervals(indices: &[usize]) -> Vec<[usize; 2]> {
    let Some((&first, remaining)) = indices.split_first() else {
        return Vec::new();
    };
    let mut intervals = Vec::new();
    let mut start = first;
    let mut previous = first;
    for &index in remaining {
        if index != previous + 1 {
            intervals.push([start, previous]);
            start = index;
        }
        previous = index;
    }
    intervals.push([start, previous]);
    intervals
}

/// Native input and numerical validation errors for tracking metrics.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum TrackingMetricsError {
    EmptyFrameOffsets,
    InvalidFirstFrameOffset {
        actual: usize,
    },
    NonMonotonicFrameOffset {
        index: usize,
        previous: usize,
        actual: usize,
    },
    FrameOffsetOutOfBounds {
        index: usize,
        value: usize,
        observation_count: usize,
    },
    FinalFrameOffset {
        expected: usize,
        actual: usize,
    },
    InputLengthOverflow {
        name: &'static str,
    },
    InputLength {
        name: &'static str,
        expected: usize,
        actual: usize,
    },
    NegativeTrackId {
        index: usize,
        track_id: i64,
    },
    InvalidStatusCode {
        index: usize,
        status_code: u8,
    },
    NonFiniteInput {
        name: &'static str,
        index: usize,
    },
    DuplicateTrackId {
        frame_index: usize,
        track_id: i64,
    },
    FrameIndexOverflow,
}

impl fmt::Display for TrackingMetricsError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::EmptyFrameOffsets => write!(formatter, "Tracking metrics require frame offsets."),
            Self::InvalidFirstFrameOffset { actual } => {
                write!(
                    formatter,
                    "Tracking metrics first frame offset must be zero; got {actual}."
                )
            }
            Self::NonMonotonicFrameOffset {
                index,
                previous,
                actual,
            } => write!(
                formatter,
                "Tracking metrics frame offset {index} must be at least {previous}; got {actual}."
            ),
            Self::FrameOffsetOutOfBounds {
                index,
                value,
                observation_count,
            } => write!(
                formatter,
                "Tracking metrics frame offset {index} is {value}; only {observation_count} observations exist."
            ),
            Self::FinalFrameOffset { expected, actual } => write!(
                formatter,
                "Tracking metrics final frame offset must be {expected}; got {actual}."
            ),
            Self::InputLengthOverflow { name } => {
                write!(formatter, "Tracking metrics {name} length overflows.")
            }
            Self::InputLength {
                name,
                expected,
                actual,
            } => write!(
                formatter,
                "Tracking metrics {name} has {actual} values; expected {expected}."
            ),
            Self::NegativeTrackId { index, track_id } => write!(
                formatter,
                "Tracking metrics track ID at observation {index} must be non-negative; got {track_id}."
            ),
            Self::InvalidStatusCode { index, status_code } => write!(
                formatter,
                "Tracking metrics status at observation {index} is invalid: {status_code}."
            ),
            Self::NonFiniteInput { name, index } => write!(
                formatter,
                "Tracking metrics {name} value {index} contains NaN or Inf."
            ),
            Self::DuplicateTrackId {
                frame_index,
                track_id,
            } => write!(
                formatter,
                "Tracking metrics frame {frame_index} repeats track ID {track_id}."
            ),
            Self::FrameIndexOverflow => {
                write!(formatter, "Tracking metrics frame index overflows.")
            }
        }
    }
}

impl std::error::Error for TrackingMetricsError {}

#[cfg(test)]
mod tests {
    use super::{TrackingMetricsInput, summarize_tracking_metrics};

    #[test]
    fn summarizes_identity_motion_and_confirmed_intervals() {
        let summary = summarize_tracking_metrics(TrackingMetricsInput {
            frame_offsets: &[0, 1, 3, 4],
            track_ids: &[2, 2, 7, 2],
            positions: &[0.0, 1.0, 0.0, 0.3, 1.4, 0.0, 1.0, 2.0, 0.0, 1.2, 1.8, 0.0],
            velocities: &[0.0, 0.0, 0.0, 0.3, 0.4, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 0.0],
            status_codes: &[0, 1, 0, 1],
            scenery_boxes: None,
            frame_index_offset: 10,
        })
        .unwrap();

        let track = &summary.tracks[0];
        assert_eq!(summary.frames_with_tracks, 3);
        assert_eq!(summary.frames_with_confirmed_tracks, 2);
        assert_eq!(track.first_frame_index, 10);
        assert_eq!(track.last_frame_index, 12);
        assert_eq!(track.confirmed_intervals, vec![[11, 12]]);
        assert!((track.path_length_m - (0.5 + 0.9_f32.hypot(0.4))).abs() < 1.0e-6);
    }
}
